File: src/test/scala/com/exasol/cloudetl/storage/StoragePropertiesTest.scala
package com.exasol.cloudetl.storage

import com.exasol.{ExaConnectionInformation, ExaMetadata}

import org.mockito.Mockito.when
import org.scalatest.BeforeAndAfterEach
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.mockito.MockitoSugar

class StoragePropertiesTest extends AnyFunSuite with BeforeAndAfterEach with MockitoSugar {

  private[this] var properties: Map[String, String] = _

  override final def beforeEach(): Unit = {
    properties = Map.empty[String, String]
    ()
  }

  test("getStoragePath returns storage path property value") {
    val path = "a/bucket/path"
    properties = Map(StorageProperties.BUCKET_PATH -> path)
    assert(BaseProperties(properties).getStoragePath() === path)
  }

  test("getStoragePath throws if storage path property is not set") {
    val thrown = intercept[IllegalArgumentException] {
      BaseProperties(properties).getStoragePath()
    }
    assert(
      thrown.getMessage === s"Please provide a value for the "
        + s"${StorageProperties.BUCKET_PATH} property!"
    )
  }

  test("getStoragePathScheme returns path scheme value") {
    val schemes = Seq("s3a", "s3", "wasbs", "adls", "file")
    schemes.foreach { scheme =>
      val path = s"$scheme://a/path"
      properties = Map(StorageProperties.BUCKET_PATH -> path)
      assert(BaseProperties(properties).getStoragePathScheme() === scheme)
    }
  }

  test("getStoragePathScheme returns path scheme with regex pattern") {
    val path = "s3a://bucket/{year=2019/month=1,year=2019/month=2}/*"
    properties = Map(StorageProperties.BUCKET_PATH -> path)
    assert(BaseProperties(properties).getStoragePathScheme() === "s3a")
  }

  test("getDeltaFormatLogStoreClassName returns storage class name for scheme") {
    val data = Map(
      "s3a" -> "org.apache.spark.sql.delta.storage.S3SingleDriverLogStore",
      "abfs" -> "org.apache.spark.sql.delta.storage.AzureLogStore",
      "abfss" -> "org.apache.spark.sql.delta.storage.AzureLogStore",
      "adl" -> "org.apache.spark.sql.delta.storage.AzureLogStore",
      "wasbs" -> "org.apache.spark.sql.delta.storage.AzureLogStore",
      "wasb" -> "org.apache.spark.sql.delta.storage.AzureLogStore",
      "file" -> "org.apache.spark.sql.delta.storage.HDFSLogStore"
    )
    data.foreach {
      case (scheme, expected) =>
        val path = s"$scheme://a/path"
        properties = Map(StorageProperties.BUCKET_PATH -> path)
        assert(BaseProperties(properties).getDeltaFormatLogStoreClassName() === expected)
    }
  }

  test("getDeltaFormatLogStoreClassName throws for unsupported storage systems") {
    Seq("gs").foreach { scheme =>
      val path = s"$scheme://a/path"
      properties = Map(StorageProperties.BUCKET_PATH -> path)
      val thrown = intercept[UnsupportedOperationException] {
        BaseProperties(properties).getDeltaFormatLogStoreClassName()
      }
      assert(thrown.getMessage.startsWith("Delta format LogStore API is not supported"))
    }
  }

  test("getFileFormat returns supported file format value") {
    properties = Map(
      StorageProperties.BUCKET_PATH -> "path",
      StorageProperties.DATA_FORMAT -> "orc"
    )
    assert(BaseProperties(properties).getFileFormat() === FileFormat.ORC)
  }

  test("getFileFormat throws if file format is not supported") {
    val fileFormat = "a-non-supported-file-format"
    properties = Map(
      StorageProperties.BUCKET_PATH -> "path",
      StorageProperties.DATA_FORMAT -> fileFormat
    )
    val thrown = intercept[IllegalArgumentException] {
      BaseProperties(properties).getFileFormat()
    }
    assert(thrown.getMessage === s"Unsupported file format $fileFormat!")
  }

  test("getParallelism returns user provided value") {
    properties = Map(StorageProperties.PARALLELISM -> "2")
    assert(BaseProperties(properties).getParallelism("default") === "2")
  }

  test("getParallelism returns default value if parallelism is not set") {
    assert(BaseProperties(properties).getParallelism("nproc()") === "nproc()")
  }

  test("isOverwrite returns true is set") {
    properties = Map(StorageProperties.OVERWRITE -> "true")
    assert(BaseProperties(properties).isOverwrite() === true)
  }

  test("isOverwrite returns default false value if it is not set") {
    assert(BaseProperties(properties).isOverwrite() === false)
  }

  final def newConnectionInformation(
    username: String,
    password: String
  ): ExaConnectionInformation =
    new ExaConnectionInformation() {
      override def getType(): ExaConnectionInformation.ConnectionType =
        ExaConnectionInformation.ConnectionType.PASSWORD
      override def getAddress(): String = ""
      override def getUser(): String = username
      override def getPassword(): String = password
    }

  test("merge returns StorageProperties with new properties") {
    properties = Map("CONNECTION_NAME" -> "connection_info")
    val metadata = mock[ExaMetadata]
    val connectionInfo = newConnectionInformation("", "KEY1=secret1==;KEY2=sec=ret2;KEY3=secret")
    when(metadata.getConnection("connection_info")).thenReturn(connectionInfo)
    val storageProperties = StorageProperties(properties, metadata).merge("")
    assert(storageProperties.getString("KEY1") === "secret1==")
    assert(storageProperties.getString("KEY2") === "sec=ret2")
    assert(storageProperties.getString("KEY3") === "secret")
  }

  test("merge returns with keyForUsername mapped to connection username") {
    properties = Map("CONNECTION_NAME" -> "connection_info")
    val metadata = mock[ExaMetadata]
    val connectionInfo = newConnectionInformation("usernameValue", "KEY1=secret1")
    when(metadata.getConnection("connection_info")).thenReturn(connectionInfo)
    val storageProperties = StorageProperties(properties, metadata).merge("usernameKey")
    assert(storageProperties.getString("usernameKey") === "usernameValue")
    assert(storageProperties.getString("KEY1") === "secret1")
  }

  test("merge returns with keyForUsername -> connection username overwritted") {
    properties = Map("CONNECTION_NAME" -> "connection_info")
    val metadata = mock[ExaMetadata]
    val connectionInfo =
      newConnectionInformation("usernameValue", "KEY1=secret1;usernameKey=newUsername")
    when(metadata.getConnection("connection_info")).thenReturn(connectionInfo)
    val storageProperties = StorageProperties(properties, metadata).merge("usernameKey")
    assert(storageProperties.getString("usernameKey") === "newUsername")
    assert(storageProperties.getString("KEY1") === "secret1")
  }

  test("merge throws if it cannot find key=value pairs in connection passoword") {
    properties = Map("CONNECTION_NAME" -> "connection_info")
    val metadata = mock[ExaMetadata]
    val connectionInfo = newConnectionInformation("", "secret1;key=value")
    when(metadata.getConnection("connection_info")).thenReturn(connectionInfo)
    val thrown = intercept[IllegalArgumentException] {
      StorageProperties(properties, metadata).merge("")
    }
    assert(thrown.getMessage === "Connection object password does not contain key=value pairs!")
  }

  test("mkString returns empty string by default") {
    val str = BaseProperties(properties).mkString()
    assert(str.isEmpty === true)
    assert(str === "")
  }

  test("mkString returns key-value properties string") {
    properties = Map("k1" -> "v1", "k3" -> "v3", "k2" -> "v2")
    val expected = "k1 -> v1;k2 -> v2;k3 -> v3"
    assert(BaseProperties(properties).mkString() === expected)
  }

  test("apply(map) returns correct StorageProperties") {
    properties = Map("a" -> "b")
    val baseProperty = BaseProperties(properties)
    assert(StorageProperties(properties) === baseProperty)
  }

  test("apply(map) with Exasol metadata returns correct StorageProperties") {
    properties = Map("k" -> "v")
    val metadata = mock[ExaMetadata]
    val expected = StorageProperties(properties, metadata)
    assert(StorageProperties(properties, metadata) === expected)
  }

  test("apply(string) throws if input string does not contain separator") {
    val thrown = intercept[IllegalArgumentException] {
      StorageProperties("")
    }
    assert(thrown.getMessage === s"The input string is not separated by ';'!")
  }

  test("apply(string) returns correct StorageProperties") {
    properties = Map("k3" -> "v3", "k2" -> "v2")
    val baseProperty = BaseProperties(properties)
    val mkStringResult = baseProperty.mkString()
    assert(StorageProperties(mkStringResult) === baseProperty)
  }

  test("apply(string) with Exasol metadata returns correct StorageProperties") {
    properties = Map("k1" -> "v1", "k2" -> "v2")
    val metadata = mock[ExaMetadata]
    val expected = StorageProperties(properties, metadata)
    val mkStringResult = BaseProperties(properties).mkString()
    assert(StorageProperties(mkStringResult, metadata) === expected)
  }

  private[this] case class BaseProperties(val params: Map[String, String])
      extends StorageProperties(params, None)

}

File: src/test/java/com/exasol/cloudetl/storage/StoragePropertiesTest.java
package com.exasol.cloudetl.storage;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.*;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.exasol.*;
import com.exasol.cloudetl.ScalaConverters;

class StoragePropertiesTest {
    private Map<String, String> properties;

    @BeforeEach
    void beforeEach() {
        this.properties = Map.of();
    }

    @Test
    void getStoragePathReturnsStoragePathPropertyValue() {
        final String path = "a/bucket/path";
        this.properties = Map.of(StorageProperties.BUCKET_PATH, path);
        assertEquals(path, baseProperties().getStoragePath());
    }

    @Test
    void getStoragePathThrowsIfStoragePathPropertyIsNotSet() {
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class,
                () -> baseProperties().getStoragePath());
        assertTrue(thrown.getMessage().startsWith("E-IEUCS-2"));
        assertTrue(thrown.getMessage()
                .contains("Failed to get value for '" + StorageProperties.BUCKET_PATH + "' property"));
    }

    @Test
    void getStoragePathSchemeReturnsPathSchemeValue() {
        for (final String scheme : List.of("s3a", "s3", "wasbs", "adls", "file")) {
            this.properties = Map.of(StorageProperties.BUCKET_PATH, scheme + "://a/path");
            assertEquals(scheme, baseProperties().getStoragePathScheme());
        }
    }

    @Test
    void getStoragePathSchemeReturnsPathSchemeWithRegexPattern() {
        this.properties = Map.of(StorageProperties.BUCKET_PATH, "s3a://bucket/{year=2019/month=1,year=2019/month=2}/*");
        assertEquals("s3a", baseProperties().getStoragePathScheme());
    }

    @Test
    void getDeltaFormatLogStoreClassNameReturnsStorageClassNameForScheme() {
        final Map<String, String> data = Map.of("s3a", "org.apache.spark.sql.delta.storage.S3SingleDriverLogStore",
                "abfs", "org.apache.spark.sql.delta.storage.AzureLogStore", "abfss",
                "org.apache.spark.sql.delta.storage.AzureLogStore", "adl",
                "org.apache.spark.sql.delta.storage.AzureLogStore", "wasbs",
                "org.apache.spark.sql.delta.storage.AzureLogStore", "wasb",
                "org.apache.spark.sql.delta.storage.AzureLogStore", "file",
                "org.apache.spark.sql.delta.storage.HDFSLogStore");
        data.forEach((scheme, expected) -> {
            this.properties = Map.of(StorageProperties.BUCKET_PATH, scheme + "://a/path");
            assertEquals(expected, baseProperties().getDeltaFormatLogStoreClassName());
        });
    }

    @Test
    void getDeltaFormatLogStoreClassNameThrowsForUnsupportedStorageSystems() {
        this.properties = Map.of(StorageProperties.BUCKET_PATH, "gs://a/path");
        final UnsupportedOperationException thrown = assertThrows(UnsupportedOperationException.class,
                () -> baseProperties().getDeltaFormatLogStoreClassName());
        assertTrue(thrown.getMessage().startsWith("E-CSE-16"));
        assertTrue(thrown.getMessage().contains("is not supported in Google Cloud Storage"));
    }

    @Test
    void getFileFormatReturnsSupportedFileFormatValue() {
        this.properties = Map.of(StorageProperties.BUCKET_PATH, "path", StorageProperties.DATA_FORMAT, "orc");
        assertEquals(FileFormat.ORC, baseProperties().getFileFormat());
    }

    @Test
    void getFileFormatThrowsIfFileFormatIsNotSupported() {
        final String fileFormat = "a-non-supported-file-format";
        this.properties = Map.of(StorageProperties.BUCKET_PATH, "path", StorageProperties.DATA_FORMAT, fileFormat);
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class,
                () -> baseProperties().getFileFormat());
        assertTrue(thrown.getMessage().startsWith("E-CSE-17"));
        assertTrue(thrown.getMessage().contains("file format '" + fileFormat + "' is not supported."));
    }

    @Test
    void getParallelismReturnsUserProvidedValue() {
        this.properties = Map.of(StorageProperties.PARALLELISM, "2");
        assertEquals("2", ScalaConverters.optionOrElse(baseProperties().getParallelism(), "default"));
    }

    @Test
    void getParallelismReturnsDefaultValueIfParallelismIsNotSet() {
        assertEquals("nproc()", ScalaConverters.optionOrElse(baseProperties().getParallelism(), "nproc()"));
    }

    @Test
    void isOverwriteReturnsTrueIsSet() {
        this.properties = Map.of(StorageProperties.OVERWRITE, "true");
        assertTrue(baseProperties().isOverwrite());
    }

    @Test
    void isOverwriteReturnsDefaultFalseValueIfItIsNotSet() {
        assertFalse(baseProperties().isOverwrite());
    }

    @Test
    void isParquetLowercaseSchemaReturnsTrueByDefault() {
        assertTrue(baseProperties().isParquetLowercaseSchema());
    }

    @Test
    void isParquetLowercaseSchemaReturnsUserSetValue() {
        this.properties = Map.of(StorageProperties.PARQUET_LOWERCASE_SCHEMA, "false");
        assertFalse(baseProperties().isParquetLowercaseSchema());
    }

    @Test
    void mergeReturnsStoragePropertiesWithNewProperties() {
        this.properties = Map.of("CONNECTION_NAME", "connection_info");
        final ExaMetadata metadata = mock(ExaMetadata.class);
        mockConnection(metadata, newConnectionInformation("", "KEY1=secret1==;KEY2=sec=ret2;KEY3=secret"));
        final StorageProperties storageProperties = new StorageProperties(this.properties, metadata).merge("");
        assertEquals("secret1==", storageProperties.getString("KEY1"));
        assertEquals("sec=ret2", storageProperties.getString("KEY2"));
        assertEquals("secret", storageProperties.getString("KEY3"));
    }

    @Test
    void mergeReturnsWithKeyForUsernameMappedToConnectionUsername() {
        this.properties = Map.of("CONNECTION_NAME", "connection_info");
        final ExaMetadata metadata = mock(ExaMetadata.class);
        mockConnection(metadata, newConnectionInformation("usernameValue", "KEY1=secret1"));
        final StorageProperties storageProperties = new StorageProperties(this.properties, metadata).merge("usernameKey");
        assertEquals("usernameValue", storageProperties.getString("usernameKey"));
        assertEquals("secret1", storageProperties.getString("KEY1"));
    }

    @Test
    void mergeReturnsWithKeyForUsernameOverwrittenByPassword() {
        this.properties = Map.of("CONNECTION_NAME", "connection_info");
        final ExaMetadata metadata = mock(ExaMetadata.class);
        mockConnection(metadata, newConnectionInformation("usernameValue", "KEY1=secret1;usernameKey=newUsername"));
        final StorageProperties storageProperties = new StorageProperties(this.properties, metadata).merge("usernameKey");
        assertEquals("newUsername", storageProperties.getString("usernameKey"));
        assertEquals("secret1", storageProperties.getString("KEY1"));
    }

    @Test
    void mergeThrowsIfItCannotFindKeyValuePairsInConnectionPassword() {
        this.properties = Map.of("CONNECTION_NAME", "connection_info");
        final ExaMetadata metadata = mock(ExaMetadata.class);
        mockConnection(metadata, newConnectionInformation("", "secret1;key=value"));
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class,
                () -> new StorageProperties(this.properties, metadata).merge(""));
        assertTrue(thrown.getMessage().contains("does not contain key-value assignment '='"));
    }

    @Test
    void mkStringReturnsEmptyStringByDefault() {
        assertEquals("", baseProperties().mkString());
    }

    @Test
    void mkStringReturnsKeyValuePropertiesString() {
        this.properties = Map.of("k1", "v1", "k3", "v3", "k2", "v2");
        assertEquals("k1 -> v1;k2 -> v2;k3 -> v3", baseProperties().mkString());
    }

    @Test
    void applyMapReturnsCorrectStorageProperties() {
        this.properties = Map.of("a", "b");
        assertEquals(baseProperties(), StorageProperties.apply(this.properties));
    }

    @Test
    void applyMapWithExasolMetadataReturnsCorrectStorageProperties() {
        this.properties = Map.of("k", "v");
        final ExaMetadata metadata = mock(ExaMetadata.class);
        assertEquals(new StorageProperties(this.properties, metadata), StorageProperties.apply(this.properties, metadata));
    }

    @Test
    void applyStringThrowsIfInputStringDoesNotContainSeparator() {
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class, () -> StorageProperties.apply(""));
        assertTrue(thrown.getMessage().startsWith("E-IEUCS-4"));
        assertTrue(thrown.getMessage().contains("does not contain key-value assignment ' -> '"));
    }

    @Test
    void applyStringReturnsCorrectStorageProperties() {
        this.properties = Map.of("k3", "v3", "k2", "v2");
        final StorageProperties baseProperty = baseProperties();
        assertEquals(baseProperty, StorageProperties.apply(baseProperty.mkString()));
    }

    @Test
    void applyStringWithExasolMetadataReturnsCorrectStorageProperties() {
        this.properties = Map.of("k1", "v1", "k2", "v2");
        final ExaMetadata metadata = mock(ExaMetadata.class);
        assertEquals(new StorageProperties(this.properties, metadata),
                StorageProperties.apply(baseProperties().mkString(), metadata));
    }

    private StorageProperties baseProperties() {
        return new StorageProperties(this.properties);
    }

    private ExaConnectionInformation newConnectionInformation(final String username, final String password) {
        return new ExaConnectionInformation() {
            @Override
            public ConnectionType getType() {
                return ConnectionType.PASSWORD;
            }

            @Override
            public String getAddress() {
                return "";
            }

            @Override
            public String getUser() {
                return username;
            }

            @Override
            public String getPassword() {
                return password;
            }
        };
    }

    private void mockConnection(final ExaMetadata metadata, final ExaConnectionInformation connectionInformation) {
        try {
            when(metadata.getConnection("connection_info")).thenReturn(connectionInformation);
        } catch (final ExaConnectionAccessException exception) {
            throw new IllegalStateException(exception);
        }
    }
}
